src/query.py

```diff
@@ -15,8 +15,62 @@
 [BedLine(chrom='chr1', chrom_start=0, chrom_end=1, name='foo'), BedLine(chrom='chr1', chrom_start=10, chrom_end=11, name='baz')]
 """
 
-from bed import BedLine
 from collections import defaultdict
+from typing import NamedTuple
+from xmlrpc.client import Boolean
+
+from bed import BedLine
+
+QueryLine = NamedTuple(
+    "QueryLine", [("chrom", str), ("chrom_start", int), ("chrom_end", int)]
+)
+
+
+def parse_query(line: str) -> QueryLine:
+    """Parse a single line query (with three columns).
+
+    >>> parse_query('chr1   1  10')
+    QueryLine(chrom='chr1', chrom_start=1, chrom_end=10)
+
+    """
+    chrom, start, end = line.split()  # split on any white-space
+    query_line = QueryLine(chrom, int(start), int(end))
+    return query_line
+
+
+def is_value_in_range(x: int, interval: tuple[int, int]):
+    """Check if value is in range (i.e. lower_bound <= x <= upper bound).
+
+    >>> is_value_in_range(5, (4, 6))
+    True
+    >>> is_value_in_range(5, (4, 5))
+    True
+
+    """
+    return interval[0] < x <= interval[1]
+
+
+def is_overlapping(x: tuple[int, int], y: tuple[int, int]) -> Boolean:
+    """Check if 2 intervals overlap.
+
+    >>> is_overlapping((1, 10), (4, 15))
+    True
+    >>> is_overlapping((15, 17), (4, 5))
+    False
+
+    """
+    return is_value_in_range(x[1], y) or is_value_in_range(y[1], x)
```

Comment:
You can do it even more simple than that:

```python
def is_overlapping(x: tuple[int,int], y: tuple[int,int]) -> bool:
    return max(x[0], y[0]) < min(x[1], y[1])
```

The idea is that `max(x[0], y[0])` is where the right-most interval starts and `min(x[1],y[1])` is where the left-most interval ends, and the range between the two points is where the overlap is. Except, if they do not overlap, then the left-most interval stops before the right-most starts.

![overlap](https://user-images.githubusercontent.com/3206538/196026561-c43da617-e5a8-4cb0-80a5-49dbdb949324.png)


You use `<` in the function if you want touching intervals to not overlap, which is what we want when we say that we exclude the last index in the range. Otherwise, you would use `<=`.

Reply:
Thanks!

Reply:
No worries. The more work you do, the more I have to find to teach you ;)